au/apply_rational_magnitude_to_integral_test.cc

```diff
@@ -0,0 +1,376 @@
+//    is `min(PM / N, TM)`.
+//
+// 3. `N > D` (all other cases).  The only step that can overflow is the multiplication with `N`.
+//    There are two conditions we must meet to avoid overflow.  First, we must avoid going over the
+//    limit of the promoted type.  Second, we must ensure that the product is within the limits of
+//    the original type after dividing by `D`.  So the maximum non-overflowing value is the smaller
+//    of `PM` and `TM * D`, divided by `N`.  (We'll have to be careful that the `TM * D` itself
+//    doesn't overflow!  We'll cap it at `PM` if it does.)
+
+enum class IsPromotable { NO, YES };
+enum class NumFitsInPromotedType { NO, YES };
+enum class DenFitsInPromotedType { NO, YES };
+struct TestSpec {
+    IsPromotable is_promotable;
+    NumFitsInPromotedType num_fits;
+    DenFitsInPromotedType den_fits;
+};
+
+template <typename T, typename MagT>
+void validate_spec(TestSpec spec) {
+    using PromotedT = PromotedType<T>;
+    const bool is_promotable = !std::is_same<T, PromotedT>::value;
+    const bool is_expected_to_be_promotable = (spec.is_promotable == IsPromotable::YES);
+    ASSERT_EQ(is_promotable, is_expected_to_be_promotable)
+        << "Expected a type that " << (is_expected_to_be_promotable ? "is" : "is not")
+        << " promotable; got a type that " << (is_promotable ? "is" : "is not");
+
+    using Num = decltype(numerator(MagT{}));
+    constexpr auto num_value_result = get_value_result<PromotedT>(Num{});
+    const bool is_num_representable = (num_value_result.outcome == MagRepresentationOutcome::OK);
+    const bool is_num_expected_to_be_representable = (spec.num_fits == NumFitsInPromotedType::YES);
+    ASSERT_EQ(is_num_representable, is_num_expected_to_be_representable)
+        << "Expected numerator " << (is_num_expected_to_be_representable ? "to be" : "not to be")
+        << " representable in promoted type; it " << (is_num_representable ? "is" : "is not");
+
+    using Den = decltype(denominator(MagT{}));
+    constexpr auto den_value_result = get_value_result<PromotedT>(Den{});
+    const bool is_den_representable = (den_value_result.outcome == MagRepresentationOutcome::OK);
+    const bool is_den_expected_to_be_representable = (spec.den_fits == DenFitsInPromotedType::YES);
+    ASSERT_EQ(is_den_representable, is_den_expected_to_be_representable)
+        << "Expected denominator " << (is_den_expected_to_be_representable ? "to be" : "not to be")
+        << " representable in promoted type; it " << (is_den_representable ? "is" : "is not");
+}
+
+template <typename T, typename MagT>
+void populate_max_non_overflowing_value(TestSpec spec, MagT, T &result_out) {
+    validate_spec<T, MagT>(spec);
+    result_out = MaxNonOverflowingValue<T, MagT>::value();
+}
+
+TEST(MaxNonOverflowingValue, AlwaysZeroIfNumCannotFitInPromotedType) {
+    // Case "1" above.
+    constexpr auto huge = pow<400>(mag<10>()) / mag<3>();
+
+    {
+        int max_int = 123;
```

Comment:
Q: It looks like the value 123 is always written over. Is there any reason for this value vs 0?

Reply:
Nope: I just needed something that is clearly not 0.  :sweat_smile: